### DISCUSSION TITLE
Examples and HowTo's

### DISCUSSION THREAD

**Discussion by Vladimir-Va** [jreyesr/n8n-nodes-carbonejs#10] [Q&A]
First of all, please accept my gratitude for your work, I'm just at the beginning of the journey and I'm full of encouragement because it worked.
I'm using n8n-nodes-carbonejs exactly as in this example:
![sample_wf 1](https://github.com/user-attachments/assets/768b7170-22b4-4766-ad2b-fef5f170ac61)
and have 3 rendered files on exit.
But what if i need to three items from Mock JSONs in one file?
I ask you to tell me, when you have a minute, why (I'm thinking from the user's point of view) in n8n-nodes-carbonejs have only one input?
Logically there should be two inputs:
1. $json context
2. $binary template

That is, just connect the template for rendering?

**Comment by jreyesr** [comment]
Hello @Vladimir-Va, and thanks for the kind words!

From this line

> But what if i need to three items from Mock JSONs in one file?

I understand that you'd like to render just _one_ Word document, that contains information of all three items (something like a list of items as a table in an invoice, or a list of people on bullet points). You could do that by using an Aggregate node beforehand, such that all the Items are grouped on a single one:

![image](https://github.com/user-attachments/assets/89ab2d82-089d-407b-bc92-6680e7a473f5)

![image](https://github.com/user-attachments/assets/aa274389-61e0-408d-90f6-3ce09a96ca65)

Note how the 100 items are converted into a single item that stores all the items under the key `posts`, as an array.

After that, only 1 item is provided to the Carbone node, which contains both JSON data:

![image](https://github.com/user-attachments/assets/d4f62f2a-d29c-4bec-a2ec-3e2e7c3777d4)

and also Binary data:

![image](https://github.com/user-attachments/assets/f2d4d88c-aa73-4c40-972b-f50d34e067f5)

Here's the template and the workflow JSON, in case you'd like to play with them:

[My_workflow.json](https://github.com/user-attachments/files/17037709/My_workflow.json)
[template2.docx](https://github.com/user-attachments/files/17037710/template2.docx)

> I ask you to tell me, when you have a minute, why (I'm thinking from the user's point of view) in n8n-nodes-carbonejs have only one input?
> Logically there should be two inputs:
> 
> 1. $json context
> 1. $binary template

Yes, that'd be an alternative: a node that has two inputs, like the Merge node, where, say, the top input should receive an item with JSON data, and the bottom input would receive an item with Binary data. However, keep in mind that N8N's workflows don't necessarily run with a single item, but may run with several items at once. For example, if you were to run the Carbone node with 3 input items, as is the case in the screenshot that you provided, that means "run the Carbone node three times, with three different templates and JSON contexts, and output three documents". This is a standard N8N convention, most nodes automatically run several times when provided with several input items, and each input item is processed separately. In other words, most nodes implicitly run something like a for-each loop, or like N8N's [Loop over Items](https://docs.n8n.io/integrations/builtin/core-nodes/n8n-nodes-base.splitinbatches/) node, which many times isn't necessary.

Therefore, considering that the Carbone node may receive several input items, having two separate inputs for JSON and Binary data would get confusing if you were to provide a different number of items to each input. For example, what if you provided three JSON items (on the top input) and two Binary items (on the bottom input)? How should the node behave? Shoud it pair the first JSON item to the first Binary item, the second to the second, and leave the third JSON item unused? Or should it pair each JSON item to each Binary item and produce six documents? 

Therefore, I chose to write the node so it has a single input, and the items that come in on that input should carry _both_ JSON and Binary data. That way, each item carries the template that it should use and also the data that should be rendered into the template. This is why the items that are inputs to the Carbone node should have _both_ JSON and binary data:

![image](https://github.com/user-attachments/assets/d9c1153e-d34f-4208-8b9b-a3ca7faba980)

Hope that helps. Don't hesitate to ask if you'd like more information.

**Comment by Vladimir-Va** [comment]
Thank you, yesterday I tried something like what you describe for one document and it didn’t work.
Today I tried what you described and it worked the first time.
Thanks, it just worked!

**Reply by jreyesr** [reply]
That's great to hear! I'll close this discussion, but feel free to reopen it if you have any more questions (if Github allows that), or open a new Discussion otherwise. I'm always glad to hear about other users of the node
